docs/dapp/sapphire/authentication.md

```diff
@@ -0,0 +1,209 @@
+
+Intra-contract calls always set `msg.sender` appropriately, if a contract calls
+another contract in a way which could reveal sensitive information, the calling
+contract must implement access control or authentication.
+
+By default all `eth_call` queries used to invoke contract functions have the
+`msg.sender` parameter set to `address(0x0)`. In contrast, authenticated calls are
+signed by a keypair and will have the `msg.sender` parameter correctly initialized
+(more on that later). Also, when a transaction is
+submitted it is signed by a keypair (thus costs gas and can make state updates)
+and the `msg.sender` will be set to the signing account.
+
+## Sapphire Wrapper
+
+The [@oasisprotocol/sapphire-paratime][sp-npm] Ethereum provider wrapper
+`sapphire.wrap` function will **automatically end-to-end encrypt calldata** when
+interacting with contracts on Sapphire, this is an easy way to ensure the
+calldata of your dApp transactions remain confidential - although the `from`,
+`to`, and `gasprice` parameters are not encrypted.
+
+[sp-npm]: https://www.npmjs.com/package/@oasisprotocol/sapphire-paratime
+
+:::tip Unauthenticated calls and Encryption
+
+Although the calls may be unauthenticated, they can still be encrypted!
+
+:::
+
+However, if the Sapphire wrapper has been attached to a signer then subsequent
+view calls via `eth_call` will request that the user signs them (e.g. a
+MetaMask popup), these are called **signed queries** meaning `msg.sender` will be
+set to the signing account and can be used for authentication or to implement
+access control. This may add friction to the end-user experience and can result
+in frequent pop-ups requesting they sign queries which wouldn't normally require
+any interaction on Transparent EVMs.
+
+Let's see how Sapphire interprets different contract calls. Suppose the
+following solidity code:
+
+```solidity
+contract Example {
+    address owner;
+    constructor () {
+        owner = msg.sender;
+    }
+    function isOwner () public view returns (bool) {
+        return msg.sender == owner;
+    }
+}
+```
+
+In the sample above, assuming we're calling from the same contract or account
+which created the contract, calling `isOwner` will return:
+
+ * `false`, for `eth_call`
+ * `false`, with `sapphire.wrap` but without an attached signer
+ * `true`, with `sapphire.wrap` and an attached signer
+ * `true`, if called via the contract which created it
+* `true`, if called via transaction
+
+## Caching Signed Queries
+
+When using signed queries the blockchain will be queried each time, however
+the Sapphire wrapper will cache signatures for signed queries with the same
+parameters to avoid asking the user to sign the same thing multiple times.
+
+Behind the scenes the signed queries use a "leash" to specify validity conditions
+so the query can only be performed within a block and account `nonce` range.
+These parameters are visible in the EIP-712 popup signed by the user. Queries
```

Comment:
If the query is cached, then no user interaction is required? Where does EIP-712 popup then come from?